client/allocrunner/groupservice_hook.go

```diff
@@ -117,10 +128,29 @@ func (h *groupServiceHook) Update(req *interfaces.RunnerUpdateRequest) error {
 	return h.consulClient.UpdateWorkload(oldWorkloadServices, newWorkloadServices)
 }
 
-func (h *groupServiceHook) Postrun() error {
+func (h *groupServiceHook) PreKill() {
 	h.mu.Lock()
 	defer h.mu.Unlock()
+
+	// If we have a shutdown delay deregister
+	// group services and then wait
+	// before continuing to kill tasks
 	h.deregister()
+	h.deregistered = true
+
+	h.logger.Debug("waiting before removing group service", "shutdown_delay", h.delay)
+	select {
+	case <-time.After(h.delay):
```

Comment:
No select required, just the receive is sufficient since there's only 1 case.  We should comment that this will block an agent from shutting down.

I think that's the right behavior for right now since we pass a TODO context into the TaskRunner.Kill methods. All of the AR.preKill/TR.Kill should run to completion and then allow the agent to exit.

The danger here is that if we shutdown our Consul heartbeat goroutine faster than a long shutdown delay (eg 30s), we could have services become unhealthy when they don't need to. Or a very long shutdown delay (eg mintues) could just make agent shutdowns very slow. I don't think concurrent alloc kills with agent shutdowns are enough of a concern to augment this behavior now, but we should at least comment where these subtle behaviors exist.

Reply:
Removed select, added a comment. Let me know if there's anymore specifics I can add to address those two concerns.